internal/controller/controllers/agent_controller.go

```diff
@@ -17,29 +17,193 @@ limitations under the License.
+
+	err := r.Get(ctx, req.NamespacedName, agent)
+	if err != nil {
+		r.Log.WithError(err).Errorf("Failed to get resource %s", req.NamespacedName)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	if agent.Spec.ClusterDeploymentName == nil {
+		err = errors.New("Cluster Deployment Name not set")
+		r.updateFailure(ctx, agent, err)
+		return ctrl.Result{}, nil
+	}
+	kubeKey := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+		Name:      agent.Spec.ClusterDeploymentName.Name,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+
+	// Retrieve clusterDeployment
+	if err = r.Get(ctx, kubeKey, clusterDeployment); err != nil {
+		errMsg := fmt.Sprintf("failed to get clusterDeployment with name %s in namespace %s",
+			agent.Spec.ClusterDeploymentName.Name, agent.Spec.ClusterDeploymentName.Namespace)
+		Requeue = false
+		clientError := true
+		if !k8serrors.IsNotFound(err) {
+			Requeue = true
+			clientError = false
+		}
+		clusterDeploymentRefErr := newKubeAPIError(errors.Wrapf(err, errMsg), clientError)
+		// Update that we failed to retrieve the clusterDeployment
+		r.updateFailure(ctx, agent, clusterDeploymentRefErr)
+		return ctrl.Result{Requeue: Requeue}, nil
+	}
+
+	// Retrieve cluster from the database
+	cluster, err := r.Installer.GetClusterByKubeKey(kubeKey)
+	if err != nil {
+		if gorm.IsRecordNotFoundError(err) {
+			Requeue = true
+			inventoryErr = common.NewApiError(http.StatusNotFound, err)
+		} else {
+			Requeue = false
+			inventoryErr = common.NewApiError(http.StatusInternalServerError, err)
+		}
+		// Update that we failed to retrieve the cluster from the database
+		r.updateFailure(ctx, agent, inventoryErr)
+		return ctrl.Result{Requeue: Requeue}, nil
+	}
+
+	var result ctrl.Result
+	// check for updates from user, compare spec and update if needed
+	result, err = r.updateIfNeeded(ctx, agent, cluster)
+	if err != nil {
+		r.updateFailure(ctx, agent, err)
+		return result, nil
+	}
+
+	conditionsv1.SetStatusCondition(&agent.Status.Conditions, conditionsv1.Condition{
+		Type:    adiiov1alpha1.AgentSyncedCondition,
+		Status:  corev1.ConditionTrue,
+		Reason:  adiiov1alpha1.AgentSyncedReason,
+		Message: adiiov1alpha1.AgentStateSynced,
+	})
+	if updateErr := r.Status().Update(ctx, agent); updateErr != nil {
+		r.Log.WithError(updateErr).Error("failed to update agent status")
+	}
+	return result, nil
+}
+
+func (r *AgentReconciler) updateIfNeeded(ctx context.Context, agent *adiiov1alpha1.Agent, c *common.Cluster) (ctrl.Result, error) {
+	spec := agent.Spec
+
+	var host *models.Host
+	for _, h := range c.Hosts {
+		if (*h.ID).String() == agent.Name {
+			host = h
+			break
+		}
+	}
+
+	if host == nil {
+		r.Log.Errorf("Host %s not found in cluster %s", agent.Name, c.Name)
+		return ctrl.Result{}, errors.New("Host not found in cluster")
+	}
+
+	clusterUpdate := false
+	params := &models.ClusterUpdateParams{}
+	if spec.Hostname != "" && spec.Hostname != host.RequestedHostname {
```

Comment:
maybe you can collect the changes and print them as part of the log in the end of the function?
"set hostname to ..." 
"set machine config pull to ..."
Can be helpful for debug no?

Reply:
We are printing the ClusterUpdateParams in UpdateClusterInternal already:
`log.Infof("update cluster %s with params: %+v", params.ClusterID, params.ClusterUpdateParams)`

Reply:
Need to check that this log is clear and not printing pointers